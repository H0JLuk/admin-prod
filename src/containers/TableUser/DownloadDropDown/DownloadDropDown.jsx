import React, { useCallback, useEffect, useRef, useState } from 'react';
import cn from 'classnames';
import { Button } from 'antd';
import { DownOutlined, UpOutlined } from '@ant-design/icons';
import { templateLink } from '../../../utils/helper';

import styles from './DownloadDropDown.module.css';

const DOWNLOAD_TEMPLATE = 'Скачать шаблон';
const DOWNLOAD_LOAD_TEMPLATE = 'Для загрузки';
const DOWNLOAD_DELETE_TEMPLATE = 'Для удаления';

const DownloadDropDown = () => {
    const [isOpen, setIsOpen] = useState(false);
    const ref = useRef(null);

    const toggleOpen = useCallback(() => setIsOpen((state) => !state), []);

    const closeWhenClickOutside = useCallback((e) => {
        if (!ref.current?.contains(e.target)) {
            setIsOpen(false);
        }
    }, []);

    useEffect(() => {
        if (isOpen) {
            document.addEventListener('click', closeWhenClickOutside);
        } else {
            document.removeEventListener('click', closeWhenClickOutside);
        }

        return () => document.removeEventListener('click', closeWhenClickOutside);
    }, [isOpen, closeWhenClickOutside]);

    return (
        <div className={ styles.dropdown } ref={ ref }>
            <Button
                className={ cn(styles.dropdownButton, { [styles.dropdownActive]: isOpen }) }
                onClick={ toggleOpen }
            >
                { DOWNLOAD_TEMPLATE } { !isOpen ? <DownOutlined /> : <UpOutlined />}
            </Button>
            <div className={ cn(styles.dropdownMenu, { [styles.open]: isOpen }) }>
                <div className={ styles.dropdownLinkWrapper }>
                    <a href={ templateLink('user-upload.xlsx') }>
                        { DOWNLOAD_LOAD_TEMPLATE }
                    </a>
                    <a href={ templateLink('user-delete.xlsx') }>
                        { DOWNLOAD_DELETE_TEMPLATE }
                    </a>
                </div>
            </div>
        </div>
    );
};

export default DownloadDropDown;
