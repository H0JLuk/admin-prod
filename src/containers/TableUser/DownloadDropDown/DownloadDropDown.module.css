.dropdownButton {
    border-radius: 20px !important;
    text-transform: uppercase;
    display: flex;
    align-items: center;
}

.dropdownButton,
.dropdownButton:hover,
.dropdownButton:focus {
    background-color: #edeeef;
    color: #414141;
    border-color: #edeeef;
}

.dropdownButton.dropdownActive {
    border-bottom-left-radius: 0 !important;
    border-bottom-right-radius: 0 !important;
}

.dropdownButton :global(.anticon.anticon-down) {
    font-size: 14px;
    width: 16px;
}

.dropdownMenu {
    position: relative;
    top: -5px;
    border-bottom-left-radius: 20px;
    border-bottom-right-radius: 20px;
    background-color: #edeeef;
}

.dropdownLinkWrapper a {
    display: block;
    font-size: 16px;
    text-transform: uppercase;
    padding: 5px 0;
    color: #919192;
}

.dropdownLinkWrapper {
    padding: 5px 20px;
}
