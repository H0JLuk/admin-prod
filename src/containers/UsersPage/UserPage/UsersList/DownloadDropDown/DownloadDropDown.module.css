.dropdown {
    position: relative;
}

.dropdownButton {
    border-radius: 2px !important;
    display: flex;
    align-items: center;
}

.dropdownButton,
.dropdownButton:hover,
.dropdownButton:focus {
    background-color: #f8f8f8;
    color: #171725;
    border-color: #f8f8f8;
}

.dropdownButton span[role='img'] {
    color: #c2c8d1;
}

.dropdownButton :global(.anticon.anticon-down) {
    font-size: 16px;
    width: 14px;
}

.dropdownMenu {
    position: absolute;
    bottom: 100%;
    overflow: hidden;
    max-height: 0;
    width: 100%;
    transition: max-height 0.3s;
    background-color: #f8f8f8;
}

.dropdownMenu.open {
    max-height: 100px;
}

.dropdownLinkWrapper {
    padding: 5px 0;
}

.dropdownLinkWrapper a {
    display: block;
    font-size: 14px;
    padding: 4px 15px;
    color: #171725;
    background-color: transparent;
    transition: background-color 0.3s;
}

.dropdownLinkWrapper a:hover {
    color: #21ba72;
}
