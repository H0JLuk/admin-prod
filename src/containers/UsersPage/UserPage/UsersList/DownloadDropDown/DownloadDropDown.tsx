import React, { memo, useCallback, useEffect, useRef, useState } from 'react';
import cn from 'classnames';
import { Button } from 'antd';
import { DownOutlined } from '@ant-design/icons';
import { templateLink } from '@utils/helper';

import styles from './DownloadDropDown.module.css';

const DOWNLOAD_TEMPLATE = 'Скачать шаблон';
const DOWNLOAD_LOAD_TEMPLATE = 'Для загрузки';
const DOWNLOAD_DELETE_TEMPLATE = 'Для удаления';

const DownloadDropDown = () => {
    const [isOpen, setIsOpen] = useState(false);
    const ref = useRef<HTMLDivElement>(null);

    const toggleOpen = () => setIsOpen(!isOpen);

    const closeWhenClickOutside = useCallback((e) => {
        if (!ref.current?.contains(e.target)) {
            setIsOpen(false);
        }
    }, []);

    useEffect(() => {
        if (isOpen) {
            document.addEventListener('click', closeWhenClickOutside);
        } else {
            document.removeEventListener('click', closeWhenClickOutside);
        }

        return () => document.removeEventListener('click', closeWhenClickOutside);
    }, [isOpen, closeWhenClickOutside]);

    return (
        <div className={styles.dropdown} ref={ref}>
            <Button
                className={cn(styles.dropdownButton, { [styles.dropdownActive]: isOpen })}
                onClick={toggleOpen}
            >
                {DOWNLOAD_TEMPLATE} <DownOutlined />
            </Button>
            <div className={cn(styles.dropdownMenu, { [styles.open]: isOpen })}>
                <div className={styles.dropdownLinkWrapper}>
                    <a href={templateLink('user-upload.csv')}>
                        {DOWNLOAD_LOAD_TEMPLATE}
                    </a>
                    <a href={templateLink('user-delete.csv')}>
                        {DOWNLOAD_DELETE_TEMPLATE}
                    </a>
                </div>
            </div>
        </div>
    );
};

export default memo(DownloadDropDown);
