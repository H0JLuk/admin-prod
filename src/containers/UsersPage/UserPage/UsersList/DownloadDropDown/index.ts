import DownloadDropDown from './DownloadDropDown';

export default DownloadDropDown;
