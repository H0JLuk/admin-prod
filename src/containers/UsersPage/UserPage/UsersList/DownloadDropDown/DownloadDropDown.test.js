import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import DownloadDropDown from './DownloadDropDown';


describe('<DownloadDropDown /> test', () => {

    const container = shallow(<DownloadDropDown />);

    it('should match the snapshot when dropdown is closed', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should match the snapshot when dropdown is open', () => {
        container.find('Button').simulate('click');
        expect(container.html()).toMatchSnapshot();
    });

    it('dropdown menu should open and close (click outside component)', async () => {
        const wrapNode = document.createElement('div');
        document.body.appendChild(wrapNode);
        const map = {};
        document.addEventListener = jest.fn((event, cb) => {
            map[event] = cb;
        });
        const event = {
            target: wrapNode,
        };
        const container = mount(<DownloadDropDown />, { attachTo: wrapNode });

        container.find('Button').simulate('click');
        expect(container.find('.dropdownMenu').hasClass('open')).toBe(true);
        expect(document.addEventListener).toHaveBeenCalled();
        await act(async () => {
            map.click(event);
        });
        container.update();
        expect(container.find('.dropdownMenu').hasClass('open')).toBe(false);

    });

    it('dropdown menu should not close (click inside component)', async () => {
        const container = mount(<DownloadDropDown />);
        const wrapNode = container.getDOMNode();
        const map ={};
        document.addEventListener = jest.fn((event, cb) => {
            map[event] = cb;
        });
        const event = {
            target: wrapNode,
        };
        container.find('Button').simulate('click');
        expect(container.find('.dropdownMenu').hasClass('open')).toBe(true);
        await act(async () => {
            map.click(event);
        });
        container.update();
        expect(container.find('.dropdownMenu').hasClass('open')).toBe(true);

    });
});
