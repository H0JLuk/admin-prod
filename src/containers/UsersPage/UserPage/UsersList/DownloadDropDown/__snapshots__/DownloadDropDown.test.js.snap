// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadDropDown /> test should match the snapshot when dropdown is closed 1`] = `"<div class=\\"dropdown\\"><button type=\\"button\\" class=\\"ant-btn dropdownButton\\"><span>Скачать шаблон </span><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></button><div class=\\"dropdownMenu\\"><div class=\\"dropdownLinkWrapper\\"><a href=\\"/templates/user-upload.csv\\">Для загрузки</a><a href=\\"/templates/user-delete.csv\\">Для удаления</a></div></div></div>"`;

exports[`<DownloadDropDown /> test should match the snapshot when dropdown is open 1`] = `"<div class=\\"dropdown\\"><button type=\\"button\\" class=\\"ant-btn dropdownButton dropdownActive\\"><span>Скачать шаблон </span><span role=\\"img\\" aria-label=\\"up\\" class=\\"anticon anticon-up\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"up\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z\\"></path></svg></span></button><div class=\\"dropdownMenu open\\"><div class=\\"dropdownLinkWrapper\\"><a href=\\"/templates/user-upload.csv\\">Для загрузки</a><a href=\\"/templates/user-delete.csv\\">Для удаления</a></div></div></div>"`;
